Fix is_hermitian to correctly handle block matrices

mathiscr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #27898 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
The function is_hermitian didn't work for block matrices like [[A]] where A is an HermitianOperator. Before this change, Matrix[[A]].is_hermitian returned False, whereas Matrix[[A]] is hermitian because A is hermitian. 

This pull request fixes this issue by updating the function _eval_is_matrix_hermitian.  Instead of checking M[i, j] - M[j, i].conjugate(), we now check M[i, j] - Dagger(M[j, i]) to compare the elements of the matrix.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
  * Fixed Matrix.is_hermitian for matrices of quantum operators.
<!-- END RELEASE NOTES -->

mathiscr committed:
matrix: Fix is_hermitian for block matrix handling

The function is_hermitian doesn't work for block matrices like [[A]] where A is an HermitianOperator. Before this commit, Matrix[[A]].is_hermitian returned False, whereas Matrix[[A]] is hermitian because A is hermitian.
This commit fixes this issue by checking if M[i, j] - Dagger(M[j, i]) is zero, which is the definition of an hermitian matrix.

oscarbenjamin commented on sympy/matrices/matrixbase.py:
```diff
@@ -1329,7 +1329,8 @@ def _eval_is_symbolic(self):
     # routines and has a different *args signature.  Make
     # sure the names don't clash by adding `_matrix_` in name.
     def _eval_is_matrix_hermitian(self, simpfunc):
-        herm = lambda i, j: simpfunc(self[i, j] - self[j, i].conjugate()).is_zero
+        from sympy.physics.quantum import Dagger
+        herm = lambda i, j: simpfunc(self[i, j] - Dagger(self[j, i])).is_zero
         return fuzzy_and(herm(i, j) for (i, j), v in self.iter_items())
```
The matrix module should not generally import from physics.

mathiscr replied:
Should I do these changes

```
     def _eval_is_matrix_hermitian(self, simpfunc):
-        from sympy.physics.quantum import Dagger
-        herm = lambda i, j: simpfunc(self[i, j] - Dagger(self[j, i])).is_zero
+        herm = lambda i, j: simpfunc(self[i, j] - (self[j, i]).transpose().conjugate()).is_zero
         return fuzzy_and(herm(i, j) for (i, j), v in self.iter_items())
```

To avoid importing from physics ?

mathiscr committed:
matrix: Changing Dagger to avoid physics import

The use of Dagger in matrixbase.py requires importing a physics module, which is a wrong import inside the matrix module.
This commit replaces the use of dagger with the functions transpose and conjugate to avoid this import.

oscarbenjamin commented:
Maybe @akhmerov can review this?

akhmerov commented on sympy/matrices/matrixbase.py:
```diff
@@ -1329,7 +1329,7 @@ def _eval_is_symbolic(self):
     # sure the names don't clash by adding `_matrix_` in name.
     def _eval_is_matrix_hermitian(self, simpfunc):
-        herm = lambda i, j: simpfunc(self[i, j] - self[j, i].conjugate()).is_zero
+        herm = lambda i, j: simpfunc(self[i, j] - self[j, i].transpose().conjugate()).is_zero
```
In https://github.com/sympy/sympy/pull/27422 we discussed that transpose shouldn't propagate to matrix entries. Therefore this might not work in general, I believe. Instead I propose using adjoint directly.
```suggestion
        herm = lambda i, j: simpfunc(self[i, j] - self[j, i].adjoint()).is_zero
```

mathiscr replied:
I agree with this change

mathiscr committed:
Replace .transpose().conjugate() with .adjoint()

Co-authored-by: Anton Akhmerov <anton.akhmerov@gmail.com>

oscarbenjamin commented:
Thanks both.
